src/theory/arith/inference_manager.h

```diff
@@ -0,0 +1,94 @@
+/*********************                                                        */
+/*! \file inference_manager.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Gereon Kremer
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Customized inference manager for the theory of nonlinear arithmetic
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__THEORY__ARITH__INFERENCE_MANAGER_H
+#define CVC4__THEORY__ARITH__INFERENCE_MANAGER_H
+
+#include <map>
+#include <vector>
+
+#include "theory/arith/arith_lemma.h"
+#include "theory/arith/arith_state.h"
+#include "theory/arith/nl/inference.h"
+#include "theory/arith/nl/nl_lemma_utils.h"
+#include "theory/inference_manager_buffered.h"
+
+namespace CVC4 {
+namespace theory {
+namespace arith {
+
+class TheoryArith;
+
+/**
+ * A specialized variant of the InferenceManagerBuffered, tailored to the
+ * arithmetic theory.
+ *
+ * It adds some convenience methods to send ArithLemma and adds a mechanism for
+ * waiting lemmas that can be flushed into the pending lemmas of the underlying
+ * buffered inference manager.
+ */
+class InferenceManager : public InferenceManagerBuffered
+{
+  using NodeSet = context::CDHashSet<Node, NodeHashFunction>;
+
+ public:
+  InferenceManager(TheoryArith& ta, ArithState& astate, ProofNodeManager* pnm);
+
+  /** Add a lemma (as pending lemma) to the underlying inference manager. */
+  void addLemma(std::shared_ptr<ArithLemma> lemma);
+  /** Add a lemma (as pending lemma) to the underlying inference manager. */
+  void addLemma(const ArithLemma& lemma);
+  /** Add a lemma (as pending lemma) to the underlying inference manager. */
+  void addLemma(const Node& lemma, nl::Inference inftype);
+
+  /** Add a lemma (as waiting lemma). */
+  void addWaitingLemma(std::shared_ptr<ArithLemma> lemma);
+  /** Add a lemma (as waiting lemma). */
+  void addWaitingLemma(const ArithLemma& lemma);
+  /** Add a lemma (as waiting lemma). */
+  void addWaitingLemma(const Node& lemma, nl::Inference inftype);
+
+  /** Flush all waiting lemmas to the underlying inference manager (as pending lemmas). */
```

Comment:
Does this move the waiting lemmas as pending lemmas? Or send them? 

Would it be easier to just have this method send all waiting lemmas?  It seems roundabout to do multiple steps of moving the lemmas around to me.

Reply:
As of now, this behaviour seems more in line with how I expect the new strategy for the nonlinear extension to work. I added a comment to make clearer what this method does.